### ISSUE TITLE
`const` feature proposal

### ISSUE THREAD

**Issue by tedbauer** [cucapra/seashell#98]
Currently, variables can't be used to determine array lengths or loop bounds. 
- For array definitions, no `let` statements can occur above them.
- For loop bounds, I think the parser only recognizes integer literals.
- Even if it were possible to write variables for loop bounds or array sizes, their type would be dynamic and rejected by the typechecker.

This is different from the old implementation, where I believe a declaration like `let x = 2;` would give `x` some "static-y" type, so `x` could be used in loop bounds or array definitions. @rachitnigam mentioned that such a strategy unnecessarily complicates the typechecking.

 In many MachSuite apps, there is heavy usage of macros to define static constants: see the original app in #97 for an example. The only option for writing such a program in Fuse is to actually write some int literal everywhere the static constant is needed, or to use the C preprocessor or some ad-hoc text-replace solution. I think it would be handy to have some feature to define static constants. Maybe it could work something like below.

## Implementation sketch
Instead of making `let` defs give variables a static type, have some separate `const` def.
The slightly updated program structure would look like this:
```
<funcdefs or record defs or const defs>
<decls>
<command>
```
A Fuse programmer could write a `const` def like this:
```
const bit<32> DENSITY = 10;
```
Here, `DENSITY` would have a `TStaticInt(10)` type, just like an int literal (I believe!). It could then be used to denote sizes in array definitions and loop bounds. It could also be used anywhere else an int literal could be, e.g. `let x = 5 + DENSITY;`. Additionally, I think `const` defs should be able to be written in terms of other `const` defined terms and integer literals:
```
const bit<32> DENSITY2 = DENSITY + 5;
```
Here, I think `DENSITY2` would get the type `TStaticInt(15)`.

**Comment by tedbauer** [comment]
One thing I forgot to consider: there would need to be some way to enforce that a variable created with a `const` def would be immutable.

**Comment by tedbauer** [comment]
Another thing: how to preserve the type information, e.g. in the `DENSITY` example, how to make sure   the `bit<32>` information is not lost.

**Comment by sampsyo** [comment]
This does indeed seem useful. For the syntax, I suggest making these exactly the same as `let`, except with `const` in there instead:

```
const DENSITY2: bit<32> = DENSITY + 5;
```

This would be the same as `let` and `const` in modern JavaScript, for example.

This is longer term, but it would be worth considering which of these constants should also be able to affect type-level parameters like array sizes and loop unrolling factors...

**Comment by rachitnigam** [comment]
An alternative proposal @sampsyo and I talked about is adding type-level natural numbers:
```
type M: Nat = 10;
type N: Nat = 5;
decl a: bit<10>[M bank N];
```

The benefit of making this type level is that this can easily be extended for the implementation of #69 where we'll have a notion of type-level bounds and type-level constants:
```
type B: Bound = ?? % 4
def foo(a: bit<32>[10 bank B])
```
which will imply that the banking factor is a factor of `4`.

**Comment by sampsyo** [comment]
One consequence of that route would be that we'd need a way to "decay" type-level numbers to term-level constant numbers to meet the `DENSITY` use case above.

**Comment by rachitnigam** [comment]
**TLDR**:
- A `const` at the language level (not the type level) seems like feature creep.
- A complicated type-level solution allowing for "decaying" would make polymorphism really hard.
  - Any solution that brings language level constructs to type level might have this problem.

Hm, that sounds non-trivial to do. I think `const` and `type Nat` should be kept separate. Adding `const` seems to re-introduce the complicated type-level computation that we used to do in this case:

```
let x = 1;
let y = 2;
let z = x + y; // z is static(3)
```

I would push against that to avoid feature creep. As a first pass, I think polymorphism shouldn't allow computation on type level stuff. Consider this program:
```
type M : Nat = ??
type N : Nat = M * M + 1;
type K : Nat = ??
def foo(a: bit<32>[N bank K]) ...
```
which would force the solver to create the non linear constraint (it is non-linear, right?): `(= (% (M * M + 1) K) 0)`.

A counterpoint might be that a separate `const` would solve this issue, but I don't want to create several binders each with slightly different semantics.

**Comment by sampsyo** [comment]
Yep, this is deceptively tricky! It seems like there are two "local minima" on opposite ends of the complexity spectrum:

1. A simple `const` modifier that doesn't grant any "special powers" like `static(n)` does—it just prevents mutation. Like `const` in JavaScript or `val` in Scala and Swift. These could of course not appear in types (e.g., they couldn't affect banking or loop unrolling factors, and they probably couldn't even be used for array sizes).
2. Type-level values. I'd argue that we would want to allow _some_ amount of type-level computation for crafting meaningful constraints, but you're right that we would want to be careful about going totally overboard. If we're not careful, we'll end up reimplementing [Liquid Haskell](https://ucsd-progsys.github.io/liquidhaskell-blog/). But in any case, these would be more flexible both in the computation that they can be involved in and the contexts they can be used in.

**Comment by rachitnigam** [comment]
Right, but if that is the semantics of `const`, i.e., it doesn't give a TStatic, it won't be useful for anything @tedbauer wants it for since any operations on it would still lead to a dynamic int.

> I'd argue that we would want to allow some amount of type-level computation for crafting meaningful constraints ...
Yup, I agree. Some variant of "linear equations over bound types" might work out fine.

**Comment by rachitnigam** [comment]
@tedbauer can we merge this with #69. The type level consts can be defined as:

```
type M: Nat = 1
```

The first pass of this will not allow for operations over these as a conservative restriction but we can add more features to as we implement polymorphism,

**Comment by tedbauer** [comment]
Sounds good!
